### PR TITLE
Ft/dropbox chunked uploads

### PR DESCRIPTION
**Pull request by birdbrained**

### CODE REVIEW

File: waterbutler/core/remote_logging.py [hunk 1]
```diff
@@ -57,7 +58,11 @@
 
     if action in ['download_file', 'download_zip']:
```
Comment (birdbrained) [inline waterbutler/core/remote_logging.py]: Use a set.

File: waterbutler/core/remote_logging.py [hunk 2]
```diff
@@ -57,7 +58,11 @@
...
         is_mfr_render = (ref_url_domain == settings.MFR_DOMAIN or
-                         settings.MFR_IDENTIFYING_HEADER in request['request']['headers'])
+                     settings.MFR_IDENTIFYING_HEADER in request['request']['headers'])
+        version = furl(request['request']['url']).args.get('version') or
+            furl(request['request']['url']).args.get('revision')
```
Comment (birdbrained) [inline waterbutler/core/remote_logging.py]: What's the difference between a version and a revision? Also maybe reword

File: waterbutler/providers/dropbox/provider.py [hunk 3]
```diff
@@ -254,22 +256,109 @@ def default_headers(self) -> dict:
...
+            data = await resp.json()
+            if resp.status == 409:
+                self.dropbox_conflict_error_handler(data, path.path)
+
+        return DropboxFileMetadata(data, self.folder), not exists
+
+    async def _chunked_upload(self,
+                              stream: streams.BaseStream,
+                              path_arg: dict) -> dict:
+        """
```
Comment (birdbrained) [inline waterbutler/providers/dropbox/provider.py]: Fixme

File: waterbutler/providers/dropbox/provider.py [hunk 4]
```diff
@@ -254,22 +256,109 @@ def default_headers(self) -> dict:
...
             'POST',
-            self._build_content_url('files', 'upload'),
+            self._build_content_url('files', 'upload_session', 'start'),
             headers={
                 'Content-Type': 'application/octet-stream',
-                'Dropbox-API-Arg': json.dumps(path_arg),
-                'Content-Length': str(stream.size),
+                'Dropbox-API-Arg': json.dumps({'close': False}),
             },
+<<<<<<< HEAD
```
Comment (birdbrained) [inline waterbutler/providers/dropbox/provider.py]: Fixme

File: waterbutler/server/api/v1/provider/__init__.py [hunk 5]
```diff
@@ -161,8 +161,8 @@ def on_finish(self):
                         'Range header: {}'.format(self.request.headers['Range']))
 
         # Don't log metadata requests.
-        if (method == 'GET' and ('meta' in self.request.query_arguments or
-                                 'revisions' in self.request.query_arguments)):
+        if method == 'GET' and ('meta' in self.request.query_arguments or
```
Comment (birdbrained) [inline waterbutler/server/api/v1/provider/__init__.py]: reword

File: tests/providers/dropbox/test_provider.py [hunk 6]
```diff
@@ -243,6 +245,54 @@ class TestCRUD:
...
+    async def test_complete_session(self, provider, stream_10_MB, provider_fixtures):
+
+        url = provider._build_content_url('files', 'upload_session', 'finish')
+
+        aiohttpretty.register_json_uri('POST',
+                                       url,
+                                       status=200,
+                                       body=provider_fixtures['file_metadata'])
+
+        upload_data = await provider._complete_session(stream_10_MB,
```
Comment (birdbrained) [inline tests/providers/dropbox/test_provider.py]: fixme

File: tests/providers/dropbox/test_provider.py [hunk 7]
```diff
@@ -243,6 +245,54 @@ class TestCRUD:
         assert metadata == expected
         assert aiohttpretty.has_call(method='POST', uri=url)
 
+    @pytest.mark.asyncio
+    @pytest.mark.aiohttpretty
+    async def test_create_upload_session(self, provider, provider_fixtures):
+
+        url = provider._build_content_url('files', 'upload_session', 'start')
+
+        aiohttpretty.register_json_uri('POST',
```
Comment (birdbrained) [inline tests/providers/dropbox/test_provider.py]: fixme

File: waterbutler/core/remote_logging.py [hunk 8]
```diff
@@ -55,9 +56,13 @@
         log_payload['metadata'] = source.serialize()
         log_payload['provider'] = log_payload['metadata']['provider']
 
-    if action in ['download_file', 'download_zip']:
+    if action in {'download_file', 'download_zip'}:
```
Comment (cslzchen) [inline waterbutler/core/remote_logging.py]: Let's skip the `[]`, `()`, `{}` refactor. We can have a ticket and discuss/groom it in advance.

File: waterbutler/core/remote_logging.py [hunk 9]
```diff
@@ -55,9 +56,13 @@
...
         is_mfr_render = (ref_url_domain == settings.MFR_DOMAIN or
-                         settings.MFR_IDENTIFYING_HEADER in request['request']['headers'])
+                     settings.MFR_IDENTIFYING_HEADER in request['request']['headers'])
+        version = (furl(request['request']['url']).args.get('version') or
+            furl(request['request']['url']).args.get('revision'))
+
+        log_payload['metadata']['version'] = version
```
Comment (cslzchen) [inline waterbutler/core/remote_logging.py]: This change is irrelevant to the ticket. Could it be fixes for other stuff that accidentally gets included?

File: waterbutler/providers/dropbox/provider.py [hunk 10]
```diff
@@ -254,22 +256,113 @@ def default_headers(self) -> dict:
         if conflict == 'replace':
             path_arg['mode'] = 'overwrite'
 
+        if stream.size > self.NONCHUNKED_UPLOAD_LIMIT:
+            data = await self._chunked_upload(stream, path_arg)
+        else:
+            resp = await self.make_request(
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: Similar to S3, how about make this `_contiguous_upload()`?

File: waterbutler/providers/dropbox/provider.py [hunk 11]
```diff
@@ -254,22 +256,113 @@ def default_headers(self) -> dict:
...
+                'Dropbox-API-Arg': json.dumps({'close': False}),
+            },
+            expects={200},
+            throws=core_exceptions.UploadError
+        )
+        return (await resp.json())['session_id']
+
+    async def _upload_parts(self, stream: streams.BaseStream, session_id: str):
+
+        for i in range(0, stream.size, self.CHUNK_SIZE):
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: Similar to S3, have an `_upload_part()` taking care of single part upload in the for loop.

File: waterbutler/server/api/v1/provider/__init__.py [hunk 12]
```diff
@@ -189,14 +195,19 @@ def _send_hook(self, action):
...
-                                       bytes_uploaded=self.bytes_uploaded,)
+        remote_logging.log_file_action(
+            action,
+            source=source,
+            destination=destination,
+            api_version='v1',
+            request=remote_logging._serialize_request(self.request),
+            bytes_downloaded=self.bytes_downloaded,
+            bytes_uploaded=self.bytes_uploaded
+        )
```
Comment (cslzchen) [inline waterbutler/server/api/v1/provider/__init__.py]: Changes in this file are irrelevant.

File: waterbutler/providers/dropbox/provider.py [hunk 13]
```diff
@@ -254,22 +256,113 @@ def default_headers(self) -> dict:
...
+                expects=(200, 409,),
+                throws=core_exceptions.UploadError,
+            )
+            data = await resp.json()
+            if resp.status == 409:
+                self.dropbox_conflict_error_handler(data, path.path)
+
+        return DropboxFileMetadata(data, self.folder), not exists
+
+    async def _chunked_upload(
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: There is no exception handling for chunked upload. What if it fails in the middle somewhere?

File: waterbutler/core/remote_logging.py (hunk 9 above)
Comment (birdbrained) [reply waterbutler/core/remote_logging.py]: This was from a merge conflic I think, so maybe its not neccesary

Comment (cslzchen) [reply waterbutler/core/remote_logging.py]: I see. Rebasing onto the latest develop branch may fix this.

File: waterbutler/core/remote_logging.py (hunk 8 above)
Comment (cslzchen) [reply waterbutler/core/remote_logging.py]: As discussed, I will revert/drop these changes.

File: waterbutler/core/remote_logging.py (hunk 9 above)
Comment (cslzchen) [reply waterbutler/core/remote_logging.py]: Remove this line.

File: waterbutler/server/api/v1/provider/__init__.py (hunk 12 above)
Comment (cslzchen) [reply waterbutler/server/api/v1/provider/__init__.py]: I will reset these irrelevant changes.

File: waterbutler/providers/dropbox/provider.py [hunk 14]
```diff
@@ -52,7 +52,9 @@ class DropboxProvider(provider.BaseProvider):
     Quirks: Dropbox paths are case-insensitive.
     """
     NAME = 'dropbox'
-    BASE_URL = provider_settings.BASE_URL
+    BASE_URL = settings.BASE_URL
+    NONCHUNKED_UPLOAD_LIMIT = 150000000  # 150 MB
+    CHUNK_SIZE = 4000000  # 4 MB
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: Make these two settings.

File: waterbutler/providers/dropbox/provider.py [hunk 15]
```diff
@@ -250,26 +252,145 @@ def default_headers(self) -> dict:
                      conflict: str='replace',
                      **kwargs) -> typing.Tuple[DropboxFileMetadata, bool]:
         path, exists = await self.handle_name_conflict(path, conflict=conflict)
-        path_arg = {"path": path.full_path}
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: Put this back. `path_arg` is used two lines below.

File: waterbutler/providers/dropbox/provider.py [hunk 16]
```diff
@@ -6,7 +6,7 @@
 from waterbutler.core.path import WaterButlerPath
 from waterbutler.core import exceptions as core_exceptions
 
-from waterbutler.providers.dropbox import settings as provider_settings
+from waterbutler.providers.dropbox import settings
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: Use `pd_settings` instead of settings to avoid shadowing and ambiguity.

File: waterbutler/providers/dropbox/provider.py [hunk 17]
```diff
@@ -52,7 +52,9 @@ class DropboxProvider(provider.BaseProvider):
     Quirks: Dropbox paths are case-insensitive.
     """
     NAME = 'dropbox'
-    BASE_URL = provider_settings.BASE_URL
+    BASE_URL = settings.BASE_URL
+    NONCHUNKED_UPLOAD_LIMIT = 150000000  # 150 MB
+    CHUNK_SIZE = 4000000  # 4 MB
 
     def __init__(self, auth, credentials, settings):
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: As mentioned above,  `settings` is the one of the arguments for each provider's `init()`method.

File: waterbutler/providers/dropbox/provider.py [hunk 18]
```diff
@@ -250,26 +252,145 @@ def default_headers(self) -> dict:
...
         if conflict == 'replace':
             path_arg['mode'] = 'overwrite'
 
+        if stream.size > self.NONCHUNKED_UPLOAD_LIMIT:
+            data = await self._chunked_upload(stream, path)
+        else:
+            data = await self._contiguous_upload(stream, path)
+        return DropboxFileMetadata(data, self.folder), not exists
+
+    async def _contiguous_upload(
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: `_contiguous_upload()` is expected to return the `data` three lines above. However, there is no return statement in the method.

File: waterbutler/providers/dropbox/provider.py [hunk 19]
```diff
@@ -250,26 +252,145 @@ def default_headers(self) -> dict:
...
+                headers={
+                    'Content-Type': 'application/octet-stream',
+                    'Dropbox-API-Arg': json.dumps({'path': path.full_path}),
+                    'Content-Length': str(stream.size),
+                },
+                data=stream,
+                expects=(200, 409,),
+                throws=core_exceptions.UploadError,
+            )
+            data = await resp.json()
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/provider.py]: I think it is here where we need to return the `data`.

File: waterbutler/providers/dropbox/settings.py [hunk 20]
```diff
@@ -4,4 +4,9 @@
 
 
 BASE_URL = config.get('BASE_URL', 'https://api.dropboxapi.com/2')
+
 BASE_CONTENT_URL = config.get('BASE_CONTENT_URL', 'https://content.dropboxapi.com/2/')
+
+CONTIGUOUS_UPLOAD_SIZE_LIMIT = int(config.get('CONTIGUOUS_UPLOAD_SIZE_LIMIT', 150000000))  # 150 MB
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/settings.py]: Should we set the limit lower so smaller files can enjoy the benefits of chunked upload?

File: waterbutler/providers/dropbox/settings.py [hunk 21]
```diff
@@ -4,4 +4,9 @@
...
+
+CHUNK_SIZE = int(config.get('CHUNK_SIZE', 4000000))  # 4 MB
```
Comment (cslzchen) [inline waterbutler/providers/dropbox/settings.py]: Chunk size can be anything up to `CONTIGUOUS_UPLOAD_SIZE_LIMIT`, which is 150 MB for now. Given that we enable `CutoffStream` for Dropbox, we can probably use a larger size.
